### ISSUE TITLE
Full support for incrementing version number in Settings.bundle

### ISSUE THREAD

**Issue by jdee** [fastlane/fastlane#5758]
### New Issue Checklist
- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
### Issue Description

This is a feature request:

It would be nice if fastlane provided full support for including the current version and build numbers in a settings bundle. I don't find any support for this in the increment_build_number or increment_version_number actions. There is a bit of support for including Settings.bundle/Root.plist in commit_version_bump (https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/commit_version_bump.rb#L93), but it requires the Settings.bundle to reside at the top of the repo. It also requires force: true, since the Root.plist is not added to expected_changes until after validation.

Ideally:
- Allow specification of the user defaults key or keys used for the version and build numbers
- Allow specification of the format used if using a single key, e.g., 1.0.2. (113)
- ~~Allow specification of the location of Settings.bundle~~ Take location of Settings.bundle from project file

At the moment, I can accomplish what I need using the plist gem. This would be a nice option for the stock actions mentioned above.

**Comment by vpolouchkine** [comment]
Thanks for the FR, keeping this open for upvotes.

**Comment by jdee** [comment]
Actually, I suppose the Settings.bundle location needs to come from the project file. I've updated the issue accordingly.

If I can find some time, I'm tempted to submit a PR for this.

**Comment by TKBurner** [comment]
@jdee That'd be super interesting!

**Comment by dombarnes** [comment]
If it helps, I've got this in my Fastfile
````
def update_root_plist(params)
    require 'plist'
    begin
      path = File.expand_path(params[:path])
      plist = Plist.parse_xml(path)
      # Set the XML position of the Version item
      plist["PreferenceSpecifiers"][0]["DefaultValue"] = params[:value]
      new_plist = plist.to_plist
      File.write(path, new_plist)
      return params[:value]
    rescue => ex
      Helper.log.error ex
      Helper.log.error "Unable to set value to plist file at '#{path}'".red
    end
  end
````
and I run 
````
update_root_plist(
      path: 'Resources/Settings.bundle/Root.plist', 
      value: friendly_version
    )
````
to update the file before running gym.

**Comment by jdee** [comment]
Thanks, @dombarnes. I have something similar using the plist gem. I did a bit of work in a fork of fastlane. The one thing I need to figure out is how to dig the location of the Root.plist out of the xcodeproj file. I'm sure there are some facilities for that in fastlane. If anyone has any pointers, feel free to share.

**Comment by jdee** [comment]
Now that I dig a little deeper, I realize that most project handling in fastlane is done by calling xcodebuild. I'm looking at options there, but I'm not sure whether I'll be able to get the location of the Root.plist file from xcodebuild. So here are some options:

- Find the right invocation of xcodebuild to extract the location from the xcodeproj file.
- Brute force it out using nokogiri or something.
- Just let the user pass it in, like the example above.

Also, I think it's probably best to make this a new action. I originally thought of modifying the IncrementBuildNumber and IncrementVersionNumber actions to include the settings bundle, but I think it might make most sense to separate this action. I'm not sure how often you'd increment both the marketing version and the build number using fastlane, but would you want it updating the settings bundle both times? Those things are currently clean wrappers around agvtool. Maybe they should just stay that way.

I think it might make more sense to add a new action, `update_settings_bundle`:

```
# If we can infer the location from the project or a default location
update_settings_bundle key: "CurrentAppVersion"

# In case the setting is not in the Root.plist
update_settings_bundle path: "Settings.bundle/About.plist", key: "CurrentAppVersion"

# Or if we can infer the location of the settings bundle
update_settings_bundle file: "About.plist", key: "CurrentAppVersion"
```

This action would take the current version info, after any increment, from the LaneContext.

The only other thing that would have to change is the CommitVersionBump action, to include the file that was changed. Maybe that could also be carried in the LaneContext, or it can be specified as an additional option to the CommitVersionBump action, like:

```
commit_version_bump settings: "Settings.bundle/Root.plist"
```

It would be nice to allow specification of the format for the output as well as separate keys for the marketing version and build number (and possibly formats for both). Maybe in the first cut, it's good enough just to support something generic like "#{version} (#{build})".

Any thoughts? I'll probably try to wrap this up in the next week or so and submit a PR.

**Comment by dombarnes** [comment]
I think a new action would make sense as it's not expected behaviour to have a version number in the settings bundle. Its not included as standard in a new project. 
I did file this issue though which relates https://github.com/fastlane/fastlane/issues/6482

**Comment by jdee** [comment]
Yes. I think I ran into that (mentioned it in the issue description). I had to put my Settings.bundle in the project root to make it work, but I'd like to fix that as part of this work, unless the original authors want to deal with it.

**Comment by bsqdev** [comment]
We update the Root.plist with the version number.  Looking forward to utilizing this proposed change.

**Comment by TKBurner** [comment]
@jdee Did you end up submitting a PR? Hope you are doing well 👍

**Comment by jdee** [comment]
@TKBurner I haven't yet. I did something for this at work and wanted to clear it with management first, just to be certain. There's not too much to it. Once I get the go-ahead, which sounds almost certain, I should be able to finish it off quickly.

**Comment by jdee** [comment]
@TKBurner It was approved, and I have a branch ready to go with basic functionality. I was about to submit a PR when I read that Fastlane doesn't accept new actions, which is what I had planned here. Do I need to set this up as a Plugin? That looks like it means a separate gem. The branch is here https://github.com/jdee/fastlane/tree/update-settings-bundle.

**Comment by jdee** [comment]
I opened a PR. Please close it if need be, and I'll set up a plugin.

**Comment by jdee** [comment]
I'll be making a plugin out of this. I'll post a link here to a new repo once that's ready, for community feedback before publishing the gem and doing the fastlane integration. That'll probably be a couple of days.

**Comment by TKBurner** [comment]
@jdee Sounds good! Hope you have a great week!

**Comment by jdee** [comment]
Apologies for the delay. I had an app in the store that was in desperate need of an update. Anyway, I finally put something together:

https://github.com/jdee/settings-bundle
https://github.com/jdee/settings-bundle/tree/master/SettingsBundleExample
https://github.com/jdee/settings-bundle/tree/master/fastlane-plugin-settings_bundle

Please use this issue for general discussion: https://github.com/jdee/settings-bundle/issues/1 or open others if you have something specific to discuss. I have a lot to do still, but you can use the plugin now. The `commit_version_bump` action will only work if your `Settings.bundle` is at the top of your project for now. A PR to fix that will follow soon.

**Comment by TKBurner** [comment]
@jdee Awesome!

**Comment by jdee** [comment]
I think the settings_bundle plugin is nearly there. I mainly want to get community feedback before it's released and maybe release it this weekend. @dombarnes Have you seen the new plugin? I'd be very interested to know how it works in your case (or any case other than apps I work on). 

Anyone with thoughts on this, please comment in that repo, on https://github.com/jdee/settings-bundle/issues/1 or by opening another issue. Thanks.

Also see the [PR](https://github.com/fastlane/fastlane/pull/6997) that fixes commit_version_bump. Thanks.

**Comment by jdee** [comment]
@dombarnes @TKBurner The `settings_bundle` plugin has been published. I think it's pretty robust, but I'm very interested in feedback of any kind. I'm sure there's some detail that could be improved. Please use issues in that repo if you have any thoughts. https://github.com/jdee/settings-bundle

There's still the matter of the PR for `commit_version_bump`. I had some further thoughts I should add to the PR. I'm not sure when I should expect that to be reviewed. But pending that fix, there is still a limitation on where you can put your `Settings.bundle` in a project, and you have to use the `force: true` option to include the settings bundle.

I'm not sure if this is the best place for this, but I'm not aware of a Fastlane discussion forum or other outlet for discussing new ideas: I've started thinking about building a new tool using `xcodeproj` and `plist` to replace `agvtool` and support both AGV and non-AGV version schemes. That tool could easily include a settings bundle update as an option. There are some problems to be solved, like the matter mentioned in jdee/settings-bundle#3. I'm curious if there would be much interest in something like that. Feel free to discuss that via issues in the settings-bundle repo if you prefer.

**Comment by jdee** [comment]
The PR for commit_version_bump has been merged, so I'm closing this issue. There's also a related PR in the docs repo (https://github.com/fastlane/docs/pull/245), but the bug has been fixed.

Incidentally, I found the fastlane-plugin-versioning, which supports non-AGV increment. I haven't looked closely, but I wasn't sure if it only updates the Info.plist files in a project or also increments CURRENT_PROJECT_VERSION, which would be a small change if it doesn't. I was also thinking about making a standalone command-line tool to replace agvtool for those lost souls who aren't using Fastlane for some reason. So I might still do something along those lines. Feel free to get in touch if you have any thoughts on the subject.
